runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/smithy/DocumentBuilder.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.smithy
+
+class DocumentBuilder internal constructor() {
+    val content: MutableMap<String, Document> = linkedMapOf()
+
+    infix fun String.to(value: Number) {
+        require(content[this] == null) { "Key $this is already registered in builder" }
+        content[this] = Document(value)
+    }
+
+    infix fun String.to(value: String) {
+        require(content[this] == null) { "Key $this is already registered in builder" }
+        content[this] = Document(value)
+    }
+
+    infix fun String.to(value: Boolean) {
+        require(content[this] == null) { "Key $this is already registered in builder" }
+        content[this] = Document(value)
+    }
+
+    infix fun String.to(value: Document?) {
+        require(content[this] == null) { "Key $this is already registered in builder" }
+        content[this] = value ?: Document.Null
+    }
+
+    class ListBuilder internal constructor() {
```

Comment:
Question: should we include an infix fun to `List<Document>`?

That would allow someone to do 

```kotlin
buildDocument {
    "foo" to listOf(1, 2, 3).map { Document(it) }
}
```

Reply:
If we provide that, would it also make sense to provide convenience mappers to specifically-typed lists like `List<Int?>`, `List<Boolean?>`, etc.? I have a gut feeling that _most_ of the time people are dealing with lists, they won't be heterogeneously-typed...they'll be (possibly-nullable) singly-typed lists.

Reply:
I would say no, given that we're adding `addAll` to the list builder.